### PR TITLE
Terminate headless processes on stop

### LINKED ISSUE #658: Terminate all processes on stop
Currently, the headless browser process that is instantiated stays alive.
When the engine is stopped, it must stop all its helpers, including its browser instance.

### PR DESCRIPTION
**Pull request by Ndpnt**
Alternative implementation of #675 
Fixes #658

### CODE REVIEW

**Comment by martinratinaud** [comment]
This looks good to me and more reliable than my implementation and we should keep yours

Two minor things though
- tests do not pass anymore (timeout problem I think when testing ssl errors)
- it's not clear that options passed to `fetch` are only for puppeteer and are not even passed to  the htmlOnly implementation, maybe we should differentiate them or put the "keepBrowerAlive" by default

TBD

**Comment by Ndpnt** [comment]
> it's not clear that options passed to fetch are only for puppeteer and are not even passed to the htmlOnly implementation, maybe we should differentiate them or put the "keepBrowerAlive" by default.

It's wanted. Consider `fetch` module as a black box, clients of this module do not have to know how it works internally. They pass options to the exposed `fetch` function and inside it we pass where it is needed. For know the only option is `keepBrowerAlive` but if we have option useful for the `HTMLOnly` fetcher we will also pass it the options object.

I hesitated to keep browser alive by default, and maybe we should do this as we do not have a use case for an auto managed browser. I am in favor of completely removing the auto browser management if we have to control headless browser manually by default.

File: scripts/validation/validate.js [hunk 1]
```diff
@@ -49,6 +51,10 @@ let servicesToValidate = args.filter(arg => !arg.startsWith('--'));
         throw new Error(`Could not find any service with id "${serviceId}"`);
       }
 
+      before(() => launchHeadlessBrowser());
+
+      after(() => closeHeadlessBrowser());
```
Comment (MattiSG) [inline scripts/validation/validate.js:56]:
```suggestion
      before(launchHeadlessBrowser);

      after(closeHeadlessBrowser);
```

File: src/app/fetcher/fullDomFetcher.js [hunk 2]
```diff
@@ -61,5 +64,21 @@ export default async function fetch(url, cssSelectors, { retry } = { retry: 0 })
...
       await page.close();
     }
+
+    if (!keepBrowserAlive && browser) {
+      await browser.close();
+    }
+  }
+}
+
+export async function launchHeadlessBrowser() {
```
Comment (MattiSG) [inline src/app/fetcher/fullDomFetcher.js:74]: Use a symmetric function name: `open` / `close` or `launch` / `stop` 🙂

**Comment by MattiSG** [comment]
As seen orally: let's just remove the automatic management option, this feature is not part of any public API 🙂

**Review by martinratinaud — APPROVED** [review]
LGTM

File: src/app/fetcher/fullDomFetcher.js [hunk 3]
```diff
@@ -63,3 +64,16 @@ export default async function fetch(url, cssSelectors, { retry } = { retry: 0 })
...
+  if (!browser) {
+    browser = await puppeteer.launch({ headless: true, args: [ '--no-sandbox', '--disable-setuid-sandbox' ] });
+  }
+}
+
+export async function stopHeadlessBrowser() {
+  if (browser) {
+    await browser.close();
+    browser = null;
+  }
```
Comment (MattiSG) [inline src/app/fetcher/fullDomFetcher.js:78]: > I like early exits 🙂 
```suggestion
  if (! browser) return;
  
  await browser.close();
  browser = null;
```

File: src/app/fetcher/fullDomFetcher.js [hunk 4]
```diff
@@ -63,3 +64,16 @@ export default async function fetch(url, cssSelectors, { retry } = { retry: 0 })
     }
   }
 }
+
+export async function launchHeadlessBrowser() {
+  if (!browser) {
+    browser = await puppeteer.launch({ headless: true, args: [ '--no-sandbox', '--disable-setuid-sandbox' ] });
```
Comment (MattiSG) [inline src/app/fetcher/fullDomFetcher.js:70]: > Multiline options would increase readability and ease diff if it changes in the future 🙂

**Review by MattiSG — APPROVED** [review]
I realize I had not fully understood that we had arbitrated for `fetch`, an abstraction layer, to expose the headless browser instance control. This seems a bit much to expose to the clients of this module, but this is not blocking 👍

**Pull request merged**